[strong] Add arity checking to functions

arv opened the pull request:
In strong mode it is illegal to call a function with too few arguments.

This introduces a new option: --strong-mode-runtime which injects
runtime checks as needed. This is of by default for strong mode at the
moment but we can revisit the default for this later.

arv commented:
@johnjbarton @rossberg-chromium

arv commented on src/codegeneration/PrependStatements.js:
```diff
@@ -37,5 +37,8 @@ export function prependStatements(statements, ...statementsToPrepend) {
     }
     transformed.push(statement);
   });
+  if (inProlog) {
```
This fixes #1199

I'll apply this change to master once done.

arv commented on src/codegeneration/RestParameterTransformer.js:
```diff
@@ -66,12 +65,4 @@ export class RestParameterTransformer extends ParameterTransformer {
     return transformed;
   }
-
-  transformConstructorType(tree) {
```
Better to keep this on the base class...

arv commented:
@Pasteuriser Would you mind taking a look since Andreas is out next week.

johnjbarton commented on src/codegeneration/LanguageModeTransformerTrait.js:
```diff
@@ -0,0 +1,141 @@
+ * @return {Function}
+ */
+export function LanguageModeTransformerTrait(superClass) {
+  return class extends superClass {
```
`ParseTreeTransformerClass` might be clearer. Or `superParseTreeTransformer` if you really lower case.

arv replied:
Done.

johnjbarton commented on src/codegeneration/LanguageModeTransformerTrait.js:
```diff
@@ -0,0 +1,141 @@
+      this.languageMode_ = SLOPPY_MODE;
+    }
+
+    updateLanguageMode_(list) {
```
Consider `statements`; it isn't completely accurate, but `list` is even less so.

johnjbarton replied:
perhaps this could be `findLanguageMode()` and return the mode. That puts all of the setters in `transformAndUpdateLanguageMode_()`

arv replied:
I actually had a findLanguageMode before... I inlined it for performance... I'll extract it again and see if I can remove some of the duplicate code we have in this area.

johnjbarton commented on src/Options.js:
```diff
@@ -156,6 +157,7 @@ addFeatureOption('generatorComprehension', EXPERIMENTAL);
 addFeatureOption('memberVariables', EXPERIMENTAL);
 addFeatureOption('require', EXPERIMENTAL);
 addFeatureOption('strongMode', EXPERIMENTAL);
+addFeatureOption('strongModeRuntime', EXPERIMENTAL);
```
With `strongModeRuntime` I was expecting a runtime function addition.  I guess you are trying express "some runtime costs".  `strongModeChecks` or `strongModeArityChecks`?

arv replied:
Maybe strongModeAsserts?

Pasteuriser replied:
Will this eventually be the flag associated with all of the dynamic aspects of strong mode? Some of these don't correspond so well to 'asserts', such as the various restrictions on strong mode functions (immutable bindings, lack of an 'arguments' object).

arv replied:
The intention was to add all features that have a runtime overhead under this option.

@johnjbarton Other things include checking that no holes are created on arrays, making class instances sealed, functions non constructable, writing to non existing properties errors etc. Some of these are assert like and others are making things more locked down.

johnjbarton commented:
Cool to see traits example.  LGTM

Pasteuriser commented on test/feature/StrongMode/ArityChecking.js:
```diff
@@ -0,0 +1,94 @@
+// Options: --strong-mode --strong-mode-runtime
+
```
It would be good to have some tests on other ways of invoking functions (Function.prototype.{call/apply/bind}, tagged template strings) and on function objects created in different ways (arrow functions, Function constructor).

arv replied:
Agreed. Will add more versions.

arv commented:
I moved some fixes out of this and merged...

I added more tests which made me realize that we cannot check the arity for arrow functions because arrow functions do not have an arguments object.

PTAL

Pasteuriser commented on src/Options.js:
```diff
@@ -156,6 +157,7 @@ addFeatureOption('generatorComprehension', EXPERIMENTAL);
 addFeatureOption('memberVariables', EXPERIMENTAL);
 addFeatureOption('require', EXPERIMENTAL);
 addFeatureOption('strongMode', EXPERIMENTAL);
+addFeatureOption('strongModeAsserts', EXPERIMENTAL);
```
Just copying and pasting my nitty comment from earlier

Will this eventually be the flag associated with all of the dynamic aspects of strong mode? Some of these don't correspond so well to 'asserts', such as the various restrictions on strong mode functions (immutable bindings, lack of an 'arguments' object etc).

I suppose these (along with strong objects) may be difficult/impossible to implement in Traceur though.

johnjbarton commented:
The description option name doesn't match the source now.
LGTM

arv commented:
> The description option name doesn't match the source now.

Not sure what that refers to?

I'm fine calling the option `strongModeAsserts` for now. We can always rename in the future.

johnjbarton commented:
Maybe you aren't using the github descriptions for commits like I am? Anyway, I meant:

> This introduces a new option: --strong-mode-runtime which injects

arv commented:
The commit looks correct to me:

https://github.com/google/traceur-compiler/pull/1902/commits

I stopped using pulley a while ago. I just merge with `--ff-only` these days.
